crypto/fipsmodule/ec/p521.c

```diff
@@ -0,0 +1,1002 @@
+//     3.   Double the accumulator 7 times. (note that doubling a point [a]P
+//          seven times results in [2^7*a]P).
+//     4.   Read from |p_pre_comp| the point corresponding to abs(s_i),
+//          negate it if s_i is negative, and add it to the accumulator.
+//
+// Note: this function is constant-time.
+static void ec_GFp_nistp521_point_mul(const EC_GROUP *group, EC_RAW_POINT *r,
+                                      const EC_RAW_POINT *p,
+                                      const EC_SCALAR *scalar) {
+
+  p521_felem res[3] = {{0}, {0}, {0}}, tmp[3] = {{0}, {0}, {0}}, ftmp;
+
+  // Table of multiples of P:  [2i + 1]P for i in [0, 63].
+  p521_felem p_pre_comp[P521_MUL_TABLE_SIZE][3];
+
+  // Set the first point in the table to P.
+  p521_from_generic(p_pre_comp[0][0], &p->X);
+  p521_from_generic(p_pre_comp[0][1], &p->Y);
+  p521_from_generic(p_pre_comp[0][2], &p->Z);
+
+  // Compute tmp = [2]P.
+  p521_point_double(tmp[0], tmp[1], tmp[2],
+                    p_pre_comp[0][0], p_pre_comp[0][1], p_pre_comp[0][2]);
+
+  // Generate the remaining 63 multiples of P.
+  for (size_t i = 1; i < P521_MUL_TABLE_SIZE; i++) {
+    p521_point_add(p_pre_comp[i][0], p_pre_comp[i][1], p_pre_comp[i][2],
+                   tmp[0], tmp[1], tmp[2], 0 /* both Jacobian */,
+                   p_pre_comp[i - 1][0],
+                   p_pre_comp[i - 1][1],
+                   p_pre_comp[i - 1][2]);
+  }
+
+  // Recode the scalar.
+  int16_t rnaf[P521_MUL_NWINDOWS] = {0};
+  p521_felem_mul_scalar_rwnaf(rnaf, scalar->bytes);
+
+  // Initialize the accumulator |res| with the table entry corresponding to
+  // the most significant digit of the recoded scalar (note that this digit
+  // can't be negative).
+  int16_t idx = rnaf[P521_MUL_NWINDOWS - 1] >> 1;
+  fiat_p521_select_point(res, idx, p_pre_comp, P521_MUL_TABLE_SIZE);
+
+  // Process the remaining digits of the scalar.
+  for (size_t i = P521_MUL_NWINDOWS - 2; i < P521_MUL_NWINDOWS - 1; i--) {
+    // Double |res| 7 times in each iteration.
+    for (size_t j = 0; j < P521_MUL_WSIZE; j++) {
+      p521_point_double(res[0], res[1], res[2], res[0], res[1], res[2]);
+    }
+
+    int16_t d = rnaf[i];
+    // is_neg = (d < 0) ? 1 : 0
+    int16_t is_neg = (d >> 7) & 1;
+    // d = abs(d)
+    d = (d ^ -is_neg) + is_neg;
+
+    idx = d >> 1;
+
+    // Select the point to add, in constant time.
+    fiat_p521_select_point(tmp, idx, p_pre_comp, P521_MUL_TABLE_SIZE);
+
+    // Negate y coordinate of the point tmp = (x, y); ftmp = -y.
+    p521_felem_opp(ftmp, tmp[1]);
+    // Conditionally select y or -y depending on the sign of the digit |d|.
+    p521_felem_cmovznz(tmp[1], is_neg, tmp[1], ftmp);
+
+    // Add the point to the accumulator |res|.
+    p521_point_add(res[0], res[1], res[2], res[0], res[1], res[2],
+                   0 /* both Jacobian */, tmp[0], tmp[1], tmp[2]);
+
+  }
+
+  p521_to_generic(&r->X, res[0]);
+  p521_to_generic(&r->Y, res[1]);
+  p521_to_generic(&r->Z, res[2]);
```

Comment:
Are these lines needed? I don't think r is used afterwards until they're repeated again at l. 675.

Reply:
good catch, leftover from debugging.